Joss paper (#816)
CExA-project/ddc: paper/paper.md

<!--
Copyright (C) The DDC development team, see below

SPDX-License-Identifier: MIT
-->

# DDC: The Discrete Domain Computation library

---
title: 'DDC: The Discrete Domain Computation library'
tags:

- C++
- HPC
- labelled arrays
- xarray

authors:

- name: Thomas Padioleau
  orcid: 0000-0001-5496-0013
  equal-contrib: true
  affiliation: 1

- name: Julien Bigot
  orcid: 0000-0002-0015-4304
  affiliation: 1

- name: Emily Bourne

>   order: 0000-0002-3469-2338

Reviewer: ```suggestion
  orcid: 0000-0002-3469-2338
```

  affiliation: 2

affiliations:

- name: Université Paris-Saclay, UVSQ, CNRS, CEA, Maison de la Simulation, 91191, Gif-sur-Yvette, France
  index: 1
- name: SCITAS, EPFL, CH-1015 Lausanne, Switzerland
  index: 2
date: 03 April 2025
bibliography: paper.bib

---

## Summary

> The Discrete Domain Computation (DDC) library is a C++ library designed to provide high-performance, strongly-typed labelled multidimensional arrays. Inspired by Python's Xarray and built on top of performance-portable libraries like Kokkos, DDC enables expressive, safe, and efficient numerical computations. It provides a coherent ecosystem to work with labelled dimensions from data structures to algorithms. Additionally, DDC extends functionality through modules such as FFT (based on Kokkos-fft), splines, and with a bridge to the PDI library. The library is actively used to modernize legacy scientific codes, such as the Fortran-based Gysela plasma simulation code.

Reviewer 2: Should we cite Gysela here?
```suggestion
The Discrete Domain Computation (DDC) library is a C++ library designed to provide high-performance, strongly-typed labelled multidimensional arrays. Inspired by Python's Xarray and built on top of performance-portable libraries like Kokkos, DDC enables expressive, safe, and efficient numerical computations. It provides a coherent ecosystem to work with labelled dimensions from data structures to algorithms. Additionally, DDC extends functionality through modules such as FFT (based on Kokkos-fft), splines, and with a bridge to the PDI library. The library is actively used to modernize legacy scientific codes, such as the Fortran-based Gysela plasma simulation code [@grandgirard2016].
```

```
@article{GRANDGIRARD201635,
title = {A 5D gyrokinetic full-f global semi-Lagrangian code for flux-driven ion turbulence simulations},
journal = {Computer Physics Communications},
volume = {207},
pages = {35-68},
year = {2016},
issn = {0010-4655},
doi = {https://doi.org/10.1016/j.cpc.2016.05.007},
url = {https://www.sciencedirect.com/science/article/pii/S0010465516301230},
author = {V. Grandgirard and J. Abiteboul and J. Bigot and T. Cartier-Michaud and N. Crouseilles and G. Dif-Pradalier and Ch. Ehrlacher and D. Esteve and X. Garbet and Ph. Ghendrih and G. Latu and M. Mehrenberger and C. Norscini and Ch. Passeron and F. Rozar and Y. Sarazin and E. Sonnendrücker and A. Strugarek and D. Zarzoso},
}
```

Author: Yes good idea

## Statement of need

The use of multidimensional arrays is widespread across various fields, particularly in scientific computing, where they serve as fundamental data containers. A primary motivation for their use is their potential to improve computational performance by leveraging problem-specific structure. For instance, when solving a partial differential equation that results in a stencil problem, computations typically achieve higher efficiency on a structured mesh compared to an unstructured mesh. This advantage primarily stems from a better usage of memory like predictable memory accesses and better cache utilization.

Many programming languages commonly used in scientific computing support multidimensional arrays in different ways. Fortran, a longstanding choice in the field, and Julia, a more recent language, both natively support these data structures. In contrast, the Python ecosystem relies on the popular NumPy library’s `numpy.Array` [@harris2020array]. Meanwhile, C++23 introduced `std::mdspan` to the standard library. This container was inspired by `Kokkos::View` from the Kokkos library which also serves as the foundation of DDC.

> Despite their importance, multidimensional arrays introduce several practical challenges. In a sense, they encourage the usage of implicits in the source code. A frequent source of errors is the inadvertent swapping of indices when accessing elements. Such errors can be difficult to detect, especially given the common convention of using single-letter variable names like `i` and `j` for indexing. Another challenge in medium to large codebases is the lack of semantic clarity in function signatures when using raw multidimensional arrays. When array dimensions carry specific meanings, this information is not explicitly represented in the source code, leaving it up to the user to ensure that dimensions are ordered correctly according to implicit expectations. For example it is quite usual to use the same index for multiple interpretations: looping over mesh cells identified by `i` and interpreting `i+1` as the face at right. Another example is slicing that removes dimensions, this operation may change the relative ordering of dimensions.

Reviewer 3: ```suggestion
Despite their importance, multidimensional arrays introduce several practical challenges. In a sense, they encourage the usage of implicits in the source code. A frequent source of errors is the inadvertent swapping of indices when accessing elements. Such errors can be difficult to detect, especially given the common convention of using single-letter variable names like `i` and `j` for indexing. Another challenge in medium to large codebases is the lack of semantic clarity in function signatures when using raw multidimensional arrays. When array dimensions carry specific meanings, this information is not explicitly represented in the source code, leaving it up to the user to ensure that dimensions are ordered correctly according to implicit expectations. For example it is quite usual to use the same index for multiple interpretations: looping over mesh cells identified by `i` and interpreting `i+1` as the face to the right. Another example is slicing that removes dimensions, this operation may change the relative ordering of dimensions.
```

> Despite their importance, multidimensional arrays introduce several practical challenges. In a sense, they encourage the usage of implicits in the source code. A frequent source of errors is the inadvertent swapping of indices when accessing elements. Such errors can be difficult to detect, especially given the common convention of using single-letter variable names like `i` and `j` for indexing. Another challenge in medium to large codebases is the lack of semantic clarity in function signatures when using raw multidimensional arrays. When array dimensions carry specific meanings, this information is not explicitly represented in the source code, leaving it up to the user to ensure that dimensions are ordered correctly according to implicit expectations. For example it is quite usual to use the same index for multiple interpretations: looping over mesh cells identified by `i` and interpreting `i+1` as the face to the right. Another example is slicing that removes dimensions, this operation may change the relative ordering of dimensions.

Author: For example: let's define a 4 dimensional array `arr` with `(X, Y, Z, T)` dimensions. If we slice it, with numpy notations, by `sub_arr = arr[0, :, 0, :]` then we have `(Y, T)`. I wanted to highlight that `arr` and `sub_arr` do not agree on the position of `Y` or `T`.

Does it make sense ?

> Despite their importance, multidimensional arrays introduce several practical challenges. In a sense, they encourage the usage of implicits in the source code. A frequent source of errors is the inadvertent swapping of indices when accessing elements. Such errors can be difficult to detect, especially given the common convention of using single-letter variable names like `i` and `j` for indexing. Another challenge in medium to large codebases is the lack of semantic clarity in function signatures when using raw multidimensional arrays. When array dimensions carry specific meanings, this information is not explicitly represented in the source code, leaving it up to the user to ensure that dimensions are ordered correctly according to implicit expectations. For example it is quite usual to use the same index for multiple interpretations: looping over mesh cells identified by `i` and interpreting `i+1` as the face to the right. Another example is slicing that removes dimensions, this operation may change the relative ordering of dimensions.

Reviewer 2: Ok I see what you mean. I'm not sure this is the "relative ordering" though.

ChatGPT suggests:
> Another example is slicing that removes dimensions, this can shift the positions of remaining dimensions, altering the correspondence between axis indices and their semantic meanings.

> Solutions have been proposed in Python and Julia to address these issues. In Python, the Xarray [@hoyer2017xarray] and Pandas [@reback2020pandas] libraries allow users to label dimensions that can then be used to perform computation. Following a similar approach, the "Discrete Domain Computation" (DDC) library aims to bring equivalent functionality to the C++ ecosystem. It uses a zero overhead abstraction approach, i.e. with labels fixed at compile-time, on top of different performant portable libraries, such as: Kokkos [@9485033], Kokkos Kernels (TODO: add a citation), Kokkos-fft (TODO: add a citation) and Ginkgo [@GinkgoJoss2020].

Reviewer 2: ```suggestion
Solutions have been proposed in Python and Julia to address these issues. In Python, the Xarray [@hoyer2017xarray] and Pandas [@reback2020pandas] libraries allow users to label dimensions that can then be used to perform computations. Following a similar approach, the "Discrete Domain Computation" (DDC) library aims to bring equivalent functionality to the C++ ecosystem. It uses a zero overhead abstraction approach, i.e. with labels fixed at compile-time, on top of different performant portable libraries, such as: Kokkos [@9485033], Kokkos Kernels (TODO: add a citation), Kokkos-fft (TODO: add a citation) and Ginkgo [@GinkgoJoss2020].
```

> Solutions have been proposed in Python and Julia to address these issues. In Python, the Xarray [@hoyer2017xarray] and Pandas [@reback2020pandas] libraries allow users to label dimensions that can then be used to perform computation. Following a similar approach, the "Discrete Domain Computation" (DDC) library aims to bring equivalent functionality to the C++ ecosystem. It uses a zero overhead abstraction approach, i.e. with labels fixed at compile-time, on top of different performant portable libraries, such as: Kokkos [@9485033], Kokkos Kernels (TODO: add a citation), Kokkos-fft (TODO: add a citation) and Ginkgo [@GinkgoJoss2020].

Reviewer 2: Reminder to fix these TODOs

## DDC Core key features

The DDC library is a C++ library designed for expressive and safe handling of multidimensional data. Its core component provides flexible data containers along with algorithms built on top of the performance portable Kokkos library. The library is fully compatible with Kokkos and does not attempt to hide it, allowing users to leverage Kokkos' full capabilities while benefiting from DDC’s strongly-typed, labelled multidimensional arrays.

### Containers

DDC offers two multidimensional containers designed over the C++ 23 multidimensional array `std::mdspan`:

- `Chunk` an owning container, i.e. it manages the lifetime of the underlying memory allocation,
- `ChunkSpan` is a non-owning container view over existing memory allocation.

### Strongly-typed labelled indices

DDC employs strongly-typed multi-indices to label dimensions and access data. It introduces two types of indices to access the container's data:

- `DiscreteVector` indices:
  - strongly-typed labelled integers,
  - provide a multidimensional array access semantics,

>   - always as fast access as raw multidimensional array,

Reviewer: This part is a bit unclear to me, maybe having an example of indexing with a `DiscreteVector` in the code example could be a good idea?

Author: Ok, @yasahi-hpc was also confused about this part. I guess it needs to be reworked. I will try to find a short example.

- `DiscreteElement` indices:
  - strongly-typed labelled keys or opaque identifiers,
  - provide an associative access semantics, as keys in a map container,
  - potentially slower access, depending on the type of set of `DiscreteElement`.

Unlike `DiscreteVector` indices, users cannot directly interpret the internal representation of `DiscreteElement` and must reason about them based solely on their relative position.

### Sets of `DiscreteElement`

> The semantics of DDC containers is to associate data to a set of `DiscreteElement` indices. Let us note that the set of all possible `DiscreteElement` has a total order that is typically established once and for all at program initialisation. Thus to be able to construct a DDC container one must provide a multidimensional set of `DiscreteElement` indices, only these indices can be later used to access the container’s data.

Reviewer 2: ```suggestion
The semantics of DDC containers associates data to a set of `DiscreteElement` indices. Let us note that the set of all possible `DiscreteElement` has a total order that is typically established once and for all at program initialization. Thus to be able to construct a DDC container one must provide a multidimensional set of `DiscreteElement` indices, only these indices can be later used to access the container’s data.
```

> The set of `DiscreteElement` is a customization point of the library. It takes the form of a Cartesian product of the different dimensions. DDC predefines the following sets:

Reviewer 2: I am not sure what is meant here

Author: What part ? The "customization point" ?

Reviewer 2: Yes, the sentence "The set of `DiscreteElement` is a customization point of the library. "

Author: So DDC provides `DiscreteDomain`, `StridedDiscreteDomain` and `SparseDiscreteDomain` but a user can provide its own domain for some specific pattern.

Reviewer 4: I guess you should clarify this. I'm not sure how much your typical reader will understand of the concept of "customization point" and what it means in this specific context.

Reviewer 2: ```suggestion

The library provides multiple ways to group `DiscreteElement`s into sets. Each set takes the form of a Cartesian product of the sets along each dimensions. DDC predefines the following sets:
```
?

- `DiscreteDomain`, a Cartesian product of intervals of `DiscreteElement` in each dimension,
- `StridedDiscreteDomain`, a Cartesian product of sets of `DiscreteElement` with a fixed step/stride in each dimension,
- `SparseDiscreteDomain`, a Cartesian product of explicitly ordered lists of `DiscreteElement` in each dimension.

> The performance of the container's data access depends on the properties of the set considered. Indeed the set is used to retrieve the position of a given multi-index relatively to the front multi-index. Thus the performance of this operation depends on the information available in the set.

Reviewer 2: ```suggestion
The performance of the container's data access depends on the properties of the set considered. Indeed the set is used to retrieve the position of a given multi-index relative to the front multi-index. Thus the performance of this operation depends on the information available in the set.
```

### Slicing

Similar to `std::mdspan`, DDC containers support slicing through the bracket operator that always returns a `ChunkSpan`. The key property is that the resulting slice’s set of `DiscreteElement` is a subset of the original set.

### Multidimensional algorithms

> Finally, DDC offers multidimensional algorithms to manipulate the containers and associated `DiscreteElement` indices. The parallel versions are based on Kokkos providing performance portability. Here is a list as of today:

Reviewer 2: Can we say as of vX.Y instead of as of today?

Author: Yes good idea

Author: done in cfb0aa81fc67fc12370d555afb390bf59a4d2a52

> - `parallel_deepcopy`, copies two `ChunkSpan` that are defined on the same set of `DiscreteElement`,

Reviewer 2: not sure about "copies two `ChunkSpan`". There are 2 arguments but I wouldn't say that both are copied

Author: Done in 00ac5e5f984eb759649dc8837ad7145258c398c5

            // my_array(idx1, idx1) = 1;
        }
    }

    // Extract a 1D view over Dim2 for each idx1
    for (ddc::DiscreteElement<Dim1> const idx1 : dom1) {
        // The following would NOT compile if sum_over_dim2 was called
        // with a `DiscreteDomain<Dim1>`, ensuring type safety.
        std::cout << sum_over_dim2(my_array[idx1]) << '\n';
    }

    return 0;
}
```

## DDC components

Built on top of DDC core, we also provide optional components as in the Xarray ecosystem. As of today we provide three components: fft, pdi and splines.

### DDC fft and splines

The FFT and splines components both operate on specialized dimensions provided by DDC: `UniformPointSampling` and `NonUniformPointSampling`.

- FFT extension: This component wraps the Kokkos-fft library to provide labelled semantics for the discrete Fourier transform (DFT). It converts data between a uniform sampling and Fourier coefficients in the Fourier basis. Depending on the transformation direction, the input array is defined over `UniformPointSampling`, while the output is defined over `PeriodicPointSampling`, which represents the Fourier dimension in DDC.

> - Splines extension: Similar to FFT, this component provides a spline transform, converting between sampled data and B-spline coefficients in a spline basis. Depending on the transformation direction, the input array is defined over `UniformPointSampling` or `NonUniformPointSampling`, while the output is defined over `UniformBSplines` or `NonUniformBSplines`. (TODO: add a citation)

Reviewer 2: What kind of citation do you want here?

Author: I had in mind either your thesis manuscript or Yuuichi's paper ?

Author: Btw, for citations do you think a github link should only appear as a link in the markdown or should be a reference in the bib ?

Reviewer 2: I usually put links in references but it is up to you. I think Yuuichi's paper should be sufficient. It is specific to this implementation so it makes sense to cite it here I think

Despite their importance, multidimensional arrays introduce several practical challenges. In a sense, they encourage the usage of implicit information in the source code. A frequent source of errors is the inadvertent swapping of indices when accessing elements. Such errors can be difficult to detect, especially given the common convention of using single-letter variable names like `i` and `j` for indexing. Another challenge in medium to large codebases is the lack of semantic clarity in function signatures when using raw multidimensional arrays. When array dimensions carry specific meanings, this information is not explicitly represented in the source code, leaving it up to the user to ensure that dimensions are ordered correctly according to implicit expectations. For example it is quite usual to use the same index for multiple interpretations: looping over mesh cells identified by `i` and interpreting `i+1` as the face to the right. Another example is slicing that removes dimensions, this can shift the positions of remaining dimensions, altering the correspondence between axis indices and their semantic meanings.

> Solutions have been proposed to address these issues. For example in Python, the Xarray [@hoyer2017xarray] library allows users to label dimensions that can then be used to perform computations. Following a similar approach, the "Discrete Domain Computation" (DDC) library aims to bring equivalent functionality to the C++ ecosystem. It uses a zero overhead abstraction approach, i.e. with labels fixed at compile-time, on top of different performant portable libraries, such as: Kokkos [@9485033], Kokkos Kernels [@rajamanickam2021kokkos], [kokkos-fft](https://github.com/kokkos/kokkos-fft) and Ginkgo [@GinkgoJoss2020]. Labelling at compile time is achieved by strongly typing dimensions, an approach similar to that used in units libraries such as mp-units [@Pusz_mp-units_2024], which strongly type quantities rather than dimensions.

Reviewer 2: > [kokkos-fft](https://github.com/kokkos/kokkos-fft) 

Strange to see a link as a citation. Why not add a `@software` citation for this?

Author: @yasahi-hpc Do you have a preference ?

Reviewer 2: @tpadioleau did you conclude on this. I strongly recommend putting the link in a citation so it is still visible if someone prints off the paper

Author: Ok I will do that tomorrow

Author: d094e2fa935ba99bd52706715c66fe972402d185

> The library is actively used to modernize the Fortran-based Gysela plasma simulation code [@Bourne_Gyselalib]. This simulation code relies heavily on high-dimensional arrays. While the data stored in the arrays logically has 7 dimensions, each dimension can have multiple representations, including Fourier, spline, Cartesian, and various curvilinear meshes. The legacy Fortran implementation, used to manipulate multi-dimensional arrays that stored slices of all the possible dimensions with very limited information about which dimensions were actually represented to enforce correctness at the API level. DDC enables a more explicit, strongly-typed representation of these arrays, ensuring at compile-time that function calls respect the expected dimensions. This reduces indexing errors and improves code maintainability, particularly in large-scale scientific software.

Reviewer 2: > While the data stored in the arrays logically has 7 dimensions,

Not sure what logic the reader is supposed to follow to get to 7 dimensions? Assume gyrokinetic = 5 + species + time ? 

```suggestion
The library is actively used to modernize the Fortran-based Gysela plasma simulation code [@Bourne_Gyselalib]. This simulation code relies heavily on high-dimensional arrays. While the data stored in the arrays has 7 dimensions, each dimension can have multiple representations, including Fourier, spline, Cartesian, and various curvilinear meshes. The legacy Fortran implementation, used to manipulate multi-dimensional arrays that stored slices of all the possible dimensions with very limited information about which dimensions were actually represented to enforce correctness at the API level. DDC enables a more explicit, strongly-typed representation of these arrays, ensuring at compile-time that function calls respect the expected dimensions. This reduces indexing errors and improves code maintainability, particularly in large-scale scientific software.
```

Author: > Not sure what logic the reader is supposed to follow to get to 7 dimensions? Assume gyrokinetic = 5 + species + time ?

Yes I think so

## DDC Core key features

The DDC library is a C++ library designed for expressive and safe handling of multidimensional data. Its core component provides flexible data containers along with algorithms built on top of the performance portable Kokkos library. The library is fully compatible with Kokkos and does not attempt to hide it, allowing users to leverage Kokkos' full capabilities while benefiting from DDC’s strongly-typed, labelled multidimensional arrays when and where it makes sense.

### Containers

DDC offers two multidimensional containers designed over the C++ 23 multidimensional array `std::mdspan`:

- `Chunk` an owning container, i.e. it manages the lifetime of the underlying memory allocation,

> - `ChunkSpan` is a non-owning container view on an existing memory allocation.

Author: I agree... this kind also breaks the example.

> - `ChunkSpan` is a non-owning container view on an existing memory allocation.

Reviewer 2: I guess there's not much we can do in this case. Ok for me

### Strongly-typed labelled indices

DDC employs strongly-typed multi-indices to label dimensions and access data. It introduces two types of multi-indices to access the container's data:

- `DiscreteVector` multi-indices:
  - strongly-typed labelled integers,
  - provide a multidimensional array access semantics,
  - always as fast access as raw multidimensional array,
- `DiscreteElement` multi-indices:

>   - strongly-typed labelled keys or opaque identifiers,

Author: > Probably not helped by the use of `x_c`, `y_b` which is usually used to denote a coordinate. From your figure I would expect:
> 
> ```c++
> DiscreteElement<X, Y> e(c, b)
> ```

In computer science they have integer loops with `x` and `y` :D

  - provide an associative access semantics, as keys in a map container,

>   - potentially slower access, depending on the type of set of `DiscreteElement`.

Author: I think it depends on the context. If you call `coordinate`, or use multiple arrays, `DiscreteElement` may be safer to be sure you refer to the same underlying mesh element. If there is no ambiguity, `DiscreteVector` may be better if performance matters.

>   - potentially slower access, depending on the type of set of `DiscreteElement`.

Reviewer 2: Created https://github.com/gyselax/gyselalibxx/issues/303

In a DDC container, `DiscreteElement` indices represent absolute positions, while `DiscreteVector` indices are always relative to the beginning of the container.

![Example of two sets of `DiscreteElement`.\label{fig:domains}](domains.pdf)

> For example consider Figure \autoref{fig:domains} that illustrates a two-dimensional data chunk with axes `X` and `Y`. Here `chunk_r` is a container defined over the red area and `chunk_b` is a slice of `chunk_r` over the blue area. Let us define

Reviewer 2: The reference does not work as you expected:
![image](https://github.com/user-attachments/assets/f32c55ee-f5b6-4e7a-a65a-4280fd6b33f7)

Author: Thanks did not catch that

> For example consider Figure \autoref{fig:domains} that illustrates a two-dimensional data chunk with axes `X` and `Y`. Here `chunk_r` is a container defined over the red area and `chunk_b` is a slice of `chunk_r` over the blue area. Let us define

Author: ```suggestion
For example consider \autoref{fig:domains} that illustrates a two-dimensional data chunk with axes `X` and `Y`. Here `chunk_r` is a container defined over the red area and `chunk_b` is a slice of `chunk_r` over the blue area. Let us define
```

The set of `DiscreteElement` is a customization point of the library. It takes the form of a Cartesian product of the different dimensions. DDC predefines the following sets:

- `DiscreteDomain`, a Cartesian product of intervals of `DiscreteElement` in each dimension,
- `StridedDiscreteDomain`, a Cartesian product of sets of `DiscreteElement` with a fixed step/stride in each dimension,
- `SparseDiscreteDomain`, a Cartesian product of explicitly ordered lists of `DiscreteElement` in each dimension.

The performance of the container's data access depends on the properties of the set considered. Indeed the set is used to retrieve the position of a given multi-index relative to the front multi-index. Thus the performance of this operation depends on the information available in the set.

### Slicing

Similar to `std::mdspan`, DDC containers support slicing through the bracket operator that always returns a `ChunkSpan`. The key property is that the resulting slice’s set of `DiscreteElement` is a subset of the original set.

### Multidimensional algorithms

> Finally, DDC offers multidimensional algorithms to manipulate the containers and associated `DiscreteElement` indices. The parallel versions are based on Kokkos providing performance portability. Here is a list as of version 0.5:

Reviewer 2: ```suggestion
Finally, DDC offers multidimensional algorithms to manipulate the containers and associated `DiscreteElement` indices. The parallel versions are based on Kokkos providing performance portability. Here is a list as of version 0.7:
```
?

    // Initialize DDC runtime (automatically finalized at the end of scope)
    ddc::ScopeGuard const ddc_scope;

    // Create a discrete domain for Dim1 with 5 elements
    ddc::DiscreteDomain<Dim1> const dom1
            = ddc::init_trivial_space<Dim1>(ddc::DiscreteVector<Dim1>(5));

    // Create a discrete domain for Dim2 with 7 elements
    ddc::DiscreteDomain<Dim2> const dom2
            = ddc::init_trivial_space<Dim2>(ddc::DiscreteVector<Dim2>(7));

    // Define a 2D discrete domain combining Dim1 and Dim2
    ddc::DiscreteDomain<Dim1, Dim2> const dom(dom1, dom2);

    // Create a 2D array (Chunk) to store integer values on the combined domain
    ddc::Chunk my_array("my_array", dom, ddc::HostAllocator<int>());

    // Iterate over the first dimension (Dim1)
    for (ddc::DiscreteElement<Dim1> const idx1 : dom1) {
        // Iterate over the second dimension (Dim2)
        for (ddc::DiscreteElement<Dim2> const idx2 : dom2) {
            // Assign the value 1 to each element in the 2D array
            my_array(idx1, idx2) = 1;

            // The following would NOT compile as my_array expects a DiscreteElement over Dim2
            // my_array(idx1, idx1) = 1;

>         }

Reviewer 2: Long line doesn't display correctly
![image](https://github.com/user-attachments/assets/b2b407f5-6666-4bfd-ada8-e0611fef3ae8)

Author: I will fix this

Reviewer 2: @tpadioleau doesn't seem to have been fixed

Author: working on this tomorrow

Reviewer 2: Confirmed that it is now on the page (although into the margin)

![image](https://github.com/user-attachments/assets/83fe2375-a002-41c9-a321-5034204e3c2a)

<!--
Copyright (C) The DDC development team, see below

SPDX-License-Identifier: MIT
-->

# DDC: The Discrete Domain Computation library

---
title: 'DDC: The Discrete Domain Computation library'
tags:

- C++
- HPC
- labelled arrays
- xarray

authors:

- name: Thomas Padioleau
  orcid: 0000-0001-5496-0013
  equal-contrib: true
  affiliation: 1

- name: Julien Bigot
  orcid: 0000-0002-0015-4304
  affiliation: 1

- name: Emily Bourne
  orcid: 0000-0002-3469-2338
  affiliation: 2

affiliations:

- name: Université Paris-Saclay, UVSQ, CNRS, CEA, Maison de la Simulation, 91191, Gif-sur-Yvette, France
  index: 1
- name: SCITAS, EPFL, CH-1015 Lausanne, Switzerland
  index: 2

> date: 03 April 2025

Reviewer 2: I assume so, but maybe it's overwritten automatically anyway. I don't see this date in the generated PDF
![image](https://github.com/user-attachments/assets/a89163c7-a34b-44b7-a3d6-06f7adb59ad8)

Solutions have been proposed in Python and Julia to address these issues. In Python, the Xarray [@hoyer2017xarray] and Pandas [@reback2020pandas] libraries allow users to label dimensions that can then be used to perform computations. Following a similar approach, the "Discrete Domain Computation" (DDC) library aims to bring equivalent functionality to the C++ ecosystem. It uses a zero overhead abstraction approach, i.e. with labels fixed at compile-time, on top of different performant portable libraries, such as: Kokkos [@9485033], Kokkos Kernels [@rajamanickam2021kokkos], [kokkos-fft](https://github.com/kokkos/kokkos-fft) and Ginkgo [@GinkgoJoss2020].

> The library is actively used to modernize the Fortran-based Gysela plasma simulation code [@Bourne_Gyselalib]. This simulation code relies heavily on high-dimensional arrays that span multiple representations, including Fourier, spline, Cartesian, and various curvilinear meshes. In the legacy Fortran implementation, these arrays were manipulated using implicit conventions, making it difficult to enforce correctness at the API level. DDC enables a more explicit, strongly-typed representation of these arrays, ensuring at compile-time that function calls respect the expected dimensions. This reduces indexing errors and improves code maintainability, particularly in large-scale scientific software.

Reviewer 4: ```suggestion
The library is actively used to modernize the Fortran-based Gysela plasma simulation code [@Bourne_Gyselalib]. This simulation code relies heavily on high-dimensional arrays. While the data stored in the arrays logically has 7 dimensions, each dimension can have multiple representations, including Fourier, spline, Cartesian, and various curvilinear meshes. The legacy Fortran implementation, used to manipulate multi-dimensional arrays that stored slices of all the possible dimensions with very limited information about which dimensions were actually represented to enforce correctness at the API level. DDC enables a more explicit, strongly-typed representation of these arrays, ensuring at compile-time that function calls respect the expected dimensions. This reduces indexing errors and improves code maintainability, particularly in large-scale scientific software.
```

## DDC Core key features

> The DDC library is a C++ library designed for expressive and safe handling of multidimensional data. Its core component provides flexible data containers along with algorithms built on top of the performance portable Kokkos library. The library is fully compatible with Kokkos and does not attempt to hide it, allowing users to leverage Kokkos' full capabilities while benefiting from DDC’s strongly-typed, labelled multidimensional arrays.

Reviewer 4: ```suggestion
The DDC library is a C++ library designed for expressive and safe handling of multidimensional data. Its core component provides flexible data containers along with algorithms built on top of the performance portable Kokkos library. The library is fully compatible with Kokkos and does not attempt to hide it, allowing users to leverage Kokkos' full capabilities while benefiting from DDC’s strongly-typed, labelled multidimensional arrays when and where it makes sense.
```

### Containers

DDC offers two multidimensional containers designed over the C++ 23 multidimensional array `std::mdspan`:

- `Chunk` an owning container, i.e. it manages the lifetime of the underlying memory allocation,
- `ChunkSpan` is a non-owning container view on an existing memory allocation.

### Strongly-typed labelled indices

DDC employs strongly-typed multi-indices to label dimensions and access data. It introduces two types of multi-indices to access the container's data:

- `DiscreteVector` multi-indices:
  - strongly-typed labelled integers,
  - provide a multidimensional array access semantics,
  - always as fast access as raw multidimensional array,
- `DiscreteElement` multi-indices:
  - strongly-typed labelled keys or opaque identifiers,
  - provide an associative access semantics, as keys in a map container,
  - potentially slower access, depending on the type of set of `DiscreteElement`.

In a DDC container, `DiscreteElement` indices represent absolute positions, while `DiscreteVector` indices are always relative to the beginning of the container.

![Example of two sets of `DiscreteElement`.\label{fig:domains}](domains.pdf)

> For example consider Figure \autoref{fig:domains} that illustrates a two-dimensional with axes `X` and `Y`. Here `chunk_r` is a container defined over the red area and `chunk_b` is a slice of `chunk_r` over the blue area. Let us define

Reviewer 4: ```suggestion
For example consider Figure \autoref{fig:domains} that illustrates a two-dimensional data chunk with axes `X` and `Y`. Here `chunk_r` is a container defined over the red area and `chunk_b` is a slice of `chunk_r` over the blue area. Let us define
```

- `DiscreteElement<X, Y> e(x_c, y_b)`,
- `DiscreteVector<X, Y> v(2, 1)`,
- `DiscreteVector<X, Y> w(0, 1)`.

In this case, the following expressions all refer to the same memory location:

- `chunk_r(e)`,
- `chunk_b(e)`,
- `chunk_r(v)`,
- `chunk_b(w)`.

> This highlights the fact that `DiscreteElement` provides a globally consistent indexing mechanism, while `DiscreteVector` is context-dependent and tied to the container’s origin.

Reviewer 4: ```suggestion
This highlights the fact that `DiscreteElement` provides a globally consistent indexing mechanism, while `DiscreteVector` is context-dependent and relative to the container’s origin.
```

    // Create a 2D array (Chunk) to store integer values on the combined domain
    ddc::Chunk my_array("my_array", dom, ddc::HostAllocator<int>());

    // Iterate over the first dimension (Dim1)
    for (ddc::DiscreteElement<Dim1> const idx1 : dom1) {
        // Iterate over the second dimension (Dim2)
        for (ddc::DiscreteElement<Dim2> const idx2 : dom2) {
            // Assign the value 1 to each element in the 2D array
            my_array(idx1, idx2) = 1;

            // The following would NOT compile as my_array expects a DiscreteElement over Dim2
            // my_array(idx1, idx1) = 1;
        }
    }

    // Extract a 1D view over Dim2 for each idx1
    for (ddc::DiscreteElement<Dim1> const idx1 : dom1) {
        // The following would NOT compile if sum_over_dim2 was called
        // with a `DiscreteDomain<Dim1>`, ensuring type safety.
        std::cout << sum_over_dim2(my_array[idx1]) << '\n';
    }

    return 0;
}
```

## DDC components

Built on top of DDC core, we also provide optional components as in the Xarray ecosystem. As of today we provide three components: fft, pdi and splines.

### DDC fft and splines

> The FFT and splines components both operate on specialized dimensions provided by DDC: `UniformPointSampling` and `NonUniformPointSampling`.

Reviewer 4: Why not introduce `UniformPointSampling` and `NonUniformPointSampling` before?

```
The FFT and splines components both operate on specialized dimensions provided by DDC: `UniformPointSampling` and `NonUniformPointSampling`.
```
